File: prob276_I-区域.py
import functools

def main():
    @functools.lru_cache(int(1e7))
    def solve(i: int, j: int, l: int, r: int, x: int, y: int) -> int:
        # N * K * M * M * 2 * 2
        w = r - l + 1
        s = sums[i][r + 1] - sums[i][l]

        if j < w:
            return -1
        if j == w:
            return s
        if i == 0 and j > w:
            return -1

        # i > 0, j > w
        ans = -1

        if (x, y) == (1, 0):
            x_, y_ = 1, 0
            for p in range(l, r + 1):
                for q in range(p, r + 1):
                    # M * M
                    res = solve(i - 1, j - w, p, q, x_, y_)
                    if res != -1 and s + res > ans:
                        ans = s + res

        if (x, y) == (0, 0):
            y_ = 0
            for x_ in range(2):
                for p in range(0, l + 1):
                    for q in range(l, r + 1):
                        res = solve(i - 1, j - w, p, q, x_, y_)
                        if res != -1 and s + res > ans:
                            ans = s + res

        if (x, y) == (1, 1):
            x_ = 1
            for y_ in range(2):
                for p in range(l, r + 1):
                    for q in range(r, M):
                        res = solve(i - 1, j - w, p, q, x_, y_)
                        if res != -1 and s + res > ans:
                            ans = s + res

        if (x, y) == (0, 1):
            for x_ in range(2):
                for y_ in range(2):
                    for p in range(0, l + 1):
                        for q in range(r, M):
                            res = solve(i - 1, j - w, p, q, x_, y_)
                            if res != -1 and s + res > ans:
                                ans = s + res

        return ans

    def get_best_decisions(i: int, j: int, l: int, r: int, x: int, y: int, ans: int) -> None:
        ans_best_decisions.append((i, j, l, r, x, y))

        w = r - l + 1
        s = sums[i][r + 1] - sums[i][l]

        if j == w:
            return

        # i > 0, j > w
        if (x, y) == (1, 0):
            x_, y_ = 1, 0
            for p in range(l, r + 1):
                for q in range(p, r + 1):
                    # M * M
                    res = solve(i - 1, j - w, p, q, x_, y_)
                    if res == ans - s:
                        get_best_decisions(i - 1, j - w, p, q, x_, y_, ans - s)
                        return

        if (x, y) == (0, 0):
            y_ = 0
            for x_ in range(2):
                for p in range(0, l + 1):
                    for q in range(l, r + 1):
                        res = solve(i - 1, j - w, p, q, x_, y_)
                        if res == ans - s:
                            get_best_decisions(i - 1, j - w, p, q, x_, y_, ans - s)
                            return

        if (x, y) == (1, 1):
            x_ = 1
            for y_ in range(2):
                for p in range(l, r + 1):
                    for q in range(r, M):
                        res = solve(i - 1, j - w, p, q, x_, y_)
                        if res == ans - s:
                            get_best_decisions(i - 1, j - w, p, q, x_, y_, ans - s)
                            return

        if (x, y) == (0, 1):
            for x_ in range(2):
                for y_ in range(2):
                    for p in range(0, l + 1):
                        for q in range(r, M):
                            res = solve(i - 1, j - w, p, q, x_, y_)
                            if res == ans - s:
                                get_best_decisions(i - 1, j - w, p, q, x_, y_, ans - s)
                                return

    N, M, K = list(map(int, input().split()))
    A = [[] for _ in range(N)]
    for i in range(N):
        A[i] = list(map(int, input().split()))


    sums = [[0 for _ in range(M + 1)] for _ in range(N)]
    for i in range(N):
        for j in range(1, M + 1):
            sums[i][j] = sums[i][j - 1] + A[i][j - 1]

    ans = -1
    start = []
    for i in range(N):
        for l in range(M):
            for r in range(l, M):
                for (x, y) in [(0, 0), (0, 1), (1, 0), (1, 1)]:
                    res = solve(i, K, l, r, x, y)
                    if res != -1 and res > ans:
                        ans = res
                        start = [i, K, l, r, x, y]

    if ans == -1:
        print("Oil : {}".format(0))
        return

    print("Oil : {}".format(ans))

    ans_best_decisions = []
    get_best_decisions(*start, ans)
    ans_best_decisions.reverse()
    for decision in ans_best_decisions:
        i = decision[0]
        l = decision[2]
        r = decision[3]
        for j in range(l, r + 1):
            print("{} {}".format(i + 1, j + 1))


if __name__ == "__main__":
    main()

File: prob276_I-区域.cpp
#include <iostream>
#include <cstring>
#include <vector>
#include <algorithm>

using namespace std;

const int INF = 0x3f3f3f3f;
const int MAXN = 15;
const int MAXM = 15;
const int MAXK = MAXN * MAXM;
int dp[MAXN][MAXK][MAXM][MAXM][2][2];
int A[MAXN][MAXM];
int sums[MAXN][MAXM + 1];
int N, M, K;

int solve(int i, int j, int l, int r, int x, int y)
{
    if(dp[i][j][l][r][x][y] != INF)
        return dp[i][j][l][r][x][y];

    int w = r - l + 1;
    int s = sums[i][r + 1] - sums[i][l];

    if(j < w)
        return -1;
    if(j == w)
         return s;
    if(i == 0 && j > w)
        return -1;

    // i > 0, j > w
    int ans = -1;
    if(x == 1 && y == 0)
    {
        int x_ = 1;
        int y_ = 0;
        for(int p = l; p <= r; ++p)
            for(int q = p; q <= r; ++q)
            {
                int res = solve(i - 1, j - w, p, q, x_, y_);
                if(res != -1 && s + res > ans)
                    ans = s + res;
            }
    }
    if(x == 0 && y == 0)
    {
        int y_ = 0;
        for(int x_ = 0; x_ <= 1; ++x_)
            for(int p = 0; p <= l; ++p)
                for(int q = l; q <= r; ++q)
                {
                    int res = solve(i - 1, j - w, p, q, x_, y_);
                    if(res != -1 && s + res > ans)
                        ans = s + res;
                }
    }
    if(x == 1 && y == 1)
    {
        int x_ = 1;
        for(int y_ = 0; y_ <= 1; ++y_)
            for(int p = l; p <= r; ++p)
                for(int q = r; q < M; ++q)
                {
                    int res = solve(i - 1, j - w, p, q, x_, y_);
                    if(res != -1 && s + res > ans)
                        ans = s + res;
                }
    }
    if(x == 0 && y == 1)
    {
        for(int x_ = 0; x_ <= 1; ++x_)
            for(int y_ = 0; y_ <= 1; ++y_)
                for(int p = 0; p <= l; ++p)
                    for(int q = r; q < M; ++q)
                    {
                        int res = solve(i - 1, j - w, p, q, x_, y_);
                        if(res != -1 && s + res > ans)
                            ans = s + res;
                    }
    }
    return dp[i][j][l][r][x][y] = ans;
}

void get_best_decisions(int i, int j, int l, int r, int x, int y, vector<vector<int>>& ans_best_decisions, int ans)
{
    ans_best_decisions.push_back({i, j, l, r, x, y});

    int w = r - l + 1;
    int s = sums[i][r + 1] - sums[i][l];

    if(j == w)
        return;

    // i > 0, j > w
    if(x == 1 && y == 0)
    {
        int x_ = 1;
        int y_ = 0;
        for(int p = l; p <= r; ++p)
            for(int q = p; q <= r; ++q)
            {
                int res = solve(i - 1, j - w, p, q, x_, y_);
                if(res == ans - s)
                {
                    get_best_decisions(i - 1, j - w, p, q, x_, y_, ans_best_decisions, ans - s);
                    return;
                }
            }
    }
    if(x == 0 && y == 0)
    {
        int y_ = 0;
        for(int x_ = 0; x_ <= 1; ++x_)
            for(int p = 0; p <= l; ++p)
                for(int q = l; q <= r; ++q)
                {
                    int res = solve(i - 1, j - w, p, q, x_, y_);
                    if(res == ans - s)
                    {
                        get_best_decisions(i - 1, j - w, p, q, x_, y_, ans_best_decisions, ans - s);
                        return;
                    }
                }
    }
    if(x == 1 && y == 1)
    {
        int x_ = 1;
        for(int y_ = 0; y_ <= 1; ++y_)
            for(int p = l; p <= r; ++p)
                for(int q = r; q < M; ++q)
                {
                    int res = solve(i - 1, j - w, p, q, x_, y_);
                    if(res == ans - s)
                    {
                        get_best_decisions(i - 1, j - w, p, q, x_, y_, ans_best_decisions, ans - s);
                        return;
                    }
                }
    }
    if(x == 0 && y == 1)
    {
        for(int x_ = 0; x_ <= 1; ++x_)
            for(int y_ = 0; y_ <= 1; ++y_)
                for(int p = 0; p <= l; ++p)
                    for(int q = r; q < M; ++q)
                    {
                        int res = solve(i - 1, j - w, p, q, x_, y_);
                        if(res == ans - s)
                        {
                            get_best_decisions(i - 1, j - w, p, q, x_, y_, ans_best_decisions, ans - s);
                            return;
                        }
                    }
    }
}

int main()
{
    memset(dp, INF, sizeof(dp));
    memset(A, -1, sizeof(A));
    memset(sums, -1, sizeof(sums));

    cin >> N >> M >> K;
    for(int i = 0; i < N; ++i)
        for(int j = 0; j < M; ++j)
            cin >> A[i][j];

    for(int i = 0; i < N; ++i)
        for(int j = 1; j <= M; ++j)
            sums[i][j] = sums[i][j - 1] + A[i][j - 1];

    int ans = -1;
    vector<int> start;
    for(int i = 0; i < N; ++i)
        for(int l = 0; l < M; ++l)
            for(int r = l; r < M; ++r)
                for(int x = 0; x <= 1; ++x)
                    for(int y = 0; y <= 1; ++y)
                    {
                        int res = solve(i, K, l, r, x, y);
                        if(res != -1 && res > ans)
                        {
                            ans = res;
                            start = {i, K, l, r, x, y};
                        }
                    }

    if(ans == -1)
        cout << "Oil : " << 0 << endl;
    else
    {
        cout << "Oil : " << ans << endl;

        vector<vector<int>> ans_best_decisions;
        get_best_decisions(start[0], start[1], start[2], start[3], start[4], start[5], ans_best_decisions, ans);
        reverse(ans_best_decisions.begin(), ans_best_decisions.end());
        for(vector<int>& decison: ans_best_decisions)
        {
            int i = decison[0];
            int l = decison[2];
            int r = decison[3];
            for(int j = l; j <= r; j++)
                cout << i + 1 << " " << j + 1 << endl;
        }
    }
}
